src/daft-plan/src/optimization/rules/push_down_projection.rs

```diff
@@ -0,0 +1,425 @@
+use std::{collections::HashSet, sync::Arc};
+
+use common_error::DaftResult;
+
+use daft_core::schema::Schema;
+use daft_dsl::Expr;
+
+use crate::{
+    ops::{Aggregate, Project, Source},
+    LogicalPlan,
+};
+
+use super::{ApplyOrder, OptimizerRule, Transformed};
+
+#[derive(Default)]
+pub struct PushDownProjection {}
+
+impl PushDownProjection {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    fn try_optimize_project(
+        &self,
+        projection: &Project,
+        plan: Arc<LogicalPlan>,
+    ) -> DaftResult<Transformed<Arc<LogicalPlan>>> {
+        let upstream_plan = &projection.input;
+        let upstream_schema = upstream_plan.schema();
+
+        // First, drop this projection if it is a no-op.
+        let projection_is_noop = {
+            let maybe_column_names = projection
+                .projection
+                .iter()
+                .map(|e| match e {
+                    Expr::Column(colname) => Some(colname.clone()),
+                    _ => None,
+                })
+                .collect::<Option<Vec<_>>>();
+
+            let all_columns_equal = maybe_column_names.map_or(false, |column_names| {
+                column_names
+                    .iter()
+                    .zip(upstream_schema.names().iter())
+                    .all(|(scol, pcol)| scol.as_ref() == pcol)
+            });
+
+            let no_unused_columns = upstream_schema.names().len() == projection.projection.len();
+
+            all_columns_equal && no_unused_columns
+        };
+        if projection_is_noop {
+            // Projection discarded but new root node has not been looked at;
+            // look at the new root node.
+            let new_plan = self
+                .try_optimize(upstream_plan.clone())?
+                .or(Transformed::Yes(upstream_plan.clone()))
+                .unwrap()
+                .clone();
+            return Ok(Transformed::Yes(new_plan));
```

Comment:
Should be able to return the result directly without unwrapping + cloning + rewrapping the `new_plan`.
```suggestion
            let new_plan = self
                .try_optimize(upstream_plan.clone())?
                .or(Transformed::Yes(upstream_plan.clone()));
            return Ok(new_plan);
```